pkg/datapath/iptables/iptables.go

```diff
@@ -326,6 +326,66 @@ func InstallRules(ifName string) error {
 				ingressSnatSrcAddrExclusion = node.GetIPv4ClusterRange().String()
 			}
 
+			egressSnatDstAddrExclusion := node.GetIPv4AllocRange().String()
+			if option.Config.Tunnel == option.TunnelDisabled {
+				egressSnatDstAddrExclusion = node.GetIPv4ClusterRange().String()
+			}
+
+			// Masquerade all egress traffic leaving the node
+			//
+			// This rule must be first as it has different exclusion criteria
+			// than the other rules in this table.
+			//
+			// The following conditions must be met:
+			// * May not leave on a cilium_ interface, this excludes all
+			//   tunnel traffic
+			// * Must originate from an IP in the local allocation range
+			// * Tunnel mode:
+			//   * May not be targeted to an IP in the local allocation
+			//     range
+			// * Non-tunnel mode:
+			//   * May not be targeted to an IP in the cluster range
+			if err := runProg("iptables", []string{
+				"-t", "nat",
+				"-A", ciliumPostNatChain,
+				"-s", node.GetIPv4AllocRange().String(),
+				"!", "-d", egressSnatDstAddrExclusion,
+				"!", "-o", "cilium_+",
+				"-m", "comment", "--comment", "cilium masquerade non-cluster",
+				"-j", "MASQUERADE"}, false); err != nil {
+				return err
+			}
+
+			// The following rules exclude traffic from the remaining rules via RETURNs.
+			// If any of these rules match, none of the remaining rules in this table
+			// are considered.
+			// Exclude traffic for other than ifName interface from the masquarade rules
+			if err := runProg("iptables", []string{
+				"-t", "nat",
+				"-A", ciliumPostNatChain,
+				"!", "-o", ifName,
+				"-m", "comment", "--comment", "exclude non-" + ifName + " traffic from masquerade",
+				"-j", "RETURN"}, false); err != nil {
+				return err
+			}
+			// Exclude crypto traffic from the masquarade rules
+			if err := runProg("iptables", []string{
+				"-t", "nat",
+				"-A", ciliumPostNatChain,
+				"-m", "mark", "--mark", matchFromIPSecEncrypt, // Don't match ipsec traffic
+				"-m", "comment", "--comment", "exclude encrypt from masquerade",
+				"-j", "RETURN"}, false); err != nil {
+				return err
+			}
+			if err := runProg("iptables", []string{
+				"-t", "nat",
+				"-A", ciliumPostNatChain,
+				"-m", "mark", "--mark", matchFromIPSecDecrypt, // Don't match ipsec traffic
+				"-m", "comment", "--comment", "exclude decrypt from masquerade",
+				"-j", "RETURN"}, false); err != nil {
```

Comment:
I've been considering adding ACCEPT rules for marked packets here. Even if we have rules in other tables/chains I don't believe we want them matching. My recollection is RETURN will continue processing later chains which may not be ideal. However, this code snippet here is a valid transformation of existing code intent so if we want to change it an easy follow up patch works.

Reply:
Changed the IPsec exclusion rules to ACCEPT, kept the non-cilium_host exclusion rule as RETURN as it is possible that the packet may need to hit some other rules in the same table.